synthesizer/src/process/execute.rs

```diff
@@ -226,4 +220,153 @@ impl<N: Network> Process<N> {
+    // - Process each transition in the `Execution` in reverse, assigning its transition ID to the corresponding function call.
+    fn construct_call_graph(&self, execution: &Execution<N>) -> Result<HashMap<N::TransitionID, Vec<N::TransitionID>>> {
+        // Metadata for each transition the execution.
+        struct TransitionMetadata<N: Network> {
+            uid: usize,
+            pid: ProgramID<N>,
+            fname: Identifier<N>,
+            tid: Option<N::TransitionID>,
+            children: Option<Vec<usize>>,
+        }
+
+        impl<N: Network> TransitionMetadata<N> {
+            fn new(counter: &mut usize, pid: ProgramID<N>, fname: Identifier<N>, tid: Option<N::TransitionID>) -> Self {
+                let uid = *counter;
+                *counter += 1;
+                Self { uid, pid, fname, tid, children: None }
+            }
+
+            fn is_complete(&self) -> bool {
+                self.tid.is_some() && self.children.is_some()
+            }
+        }
+
+        // A helper function to update the call graph, given transition metadata.
+        let update_call_graph = |metadata: TransitionMetadata<N>,
+                                 call_graph: &mut HashMap<N::TransitionID, Vec<N::TransitionID>>,
+                                 uid_to_tid: &mut HashMap<usize, N::TransitionID>|
+         -> Result<()> {
+            // Check that the transition metadata is complete.
+            ensure!(metadata.is_complete(), "The transition metadata is incomplete");
+            // Update the UID to TID mapping.
+            uid_to_tid.insert(metadata.uid, metadata.tid.unwrap());
+            // Update the call graph.
+            call_graph.insert(
+                metadata.tid.unwrap(),
+                metadata.children.unwrap().into_iter().map(|uid| uid_to_tid[&uid]).collect::<Vec<_>>(),
+            );
+            Ok(())
+        };
+
+        // Initialize a call graph, which is a map of transition IDs to the transition IDs it calls.
+        let mut call_graph = HashMap::new();
+        // Initialize a mapping from UIDs to transition IDs.
+        let mut uid_to_tid = HashMap::new();
+
+        // Initialize a stack to track transition metadata, while traversing the call graph.
+        let mut traversal_stack: Vec<TransitionMetadata<N>> = Vec::new();
+        // Initialize a counter to provide unique IDs for each transition.
+        let mut counter = 0;
+
+        // Iterate over each transition in reverse post-order, and populate the call graph.
+        for transition in execution.transitions().rev() {
+            // If the stack has complete metadata entries, then remove and add them to the call graph.
+            while let Some(metadata) = traversal_stack.last() {
+                if metadata.is_complete() {
+                    update_call_graph(traversal_stack.pop().unwrap(), &mut call_graph, &mut uid_to_tid)?;
+                } else {
+                    break;
+                }
+            }
```

Comment:
This can be moved to the end of the `for` loop. Then the last `while` loop at the end of this function can be removed (and an `ensure!(traversal_stack.is_empty, "...")` can be added at the end of the function).